python/cuda_cccl/cuda/cccl/parallel/experimental/algorithms/_reduce.py

```diff
@@ -134,3 +134,17 @@ def reduce_into(
         A callable object that can be used to perform the reduction
     """
     return _Reduce(d_in, d_out, op, h_init)
+
+
+def reduce_into(
+    d_in: DeviceArrayLike | IteratorBase,
+    d_out: DeviceArrayLike,
```

Comment:
Sorry I might have missed something. For single-phase APIs, shouldn't we allocate the output array for users and return it to them? This does not seem pythonic to me? Extra inputs means extra consistency checks (ex: what if `d_in.dtype` != `d_out.dtype`?) which translates to extra overhead.

Usually if we want user-provided output, we should add an optional `out=` kw-arg.

Reply:
The output need not be an array - it can be an iterator. Further, depending on the op, the output can be of a different type than `d_in` that we can't infer trivially.

Reply:
I must have missed that, do you have an example for `d_out` being an iterator?

Reply:
Yes, for example https://github.com/NVIDIA/cccl/blob/7f19fd8e2652b5fae0baf58bfee2e72b77aaad4c/python/cuda_cccl/tests/parallel/test_scan_api.py#L92.

Another common use case, but one we haven't exposed yet, is https://nvidia.github.io/cccl/thrust/api/classthrust_1_1transform__output__iterator.html

Reply:
@shwina if `d_out` can be an iterator, shouldn't we add it to the type annotation? We did this to `d_in`:
```suggestion
    d_out: DeviceArrayLike | IteratorBase,
```
(same for other algorithms)

Reply:
When we do extend this and other algorithms to handle iterators as outputs we can make this change, but it's not supported for reduce (and some other algorithms) today.